File: daySeven/phoneBook/python/phone_book.py
# phone_book.py

numbers = []

def add_contact():
    while True:
        first_name = input("Please Enter First Name: ").strip()

        duplicate_first_name = any(user['first_name'].lower() == first_name.lower() for user in numbers)
        if duplicate_first_name:
            print(f"Error: User with the first name '{first_name}' already exists. Please enter a different name.")
        else:
            break

    while True:
        last_name = input("Please Enter Last Name: ").strip()

        duplicate_last_name = any(user['last_name'].lower() == last_name.lower() for user in numbers)
        if duplicate_last_name:
            print(f"Error: User with the last name '{last_name}' already exists. Please enter a different name.")
        else:
            break

    while True:
        phone_number = input("Enter Phone Number: ").strip()

        duplicate_phone_number = any(user['phone_number'] == phone_number for user in numbers)
        if duplicate_phone_number:
            print(f"Error: Phone number '{phone_number}' already exists. Please enter a different phone number.")
        else:
            break

    contact = {
        'first_name': first_name,
        'last_name': last_name,
        'phone_number': phone_number
    }
    numbers.append(contact)
    print("Contact has been added successfully!\n")


def remove_contact():
    if not numbers:
        print("No contacts found.\n")
        return

    remove_phone_number = input("Enter the phone number you want to remove: ").strip()

    for i, contact in enumerate(numbers):
        if contact['phone_number'] == remove_phone_number:
            del numbers[i]
            print("Contact has been removed successfully.")
            return

    print("Contact not found!!")


def find_contact_by_phone_number():
    if not numbers:
        print("No contacts found.\n")
        return

    find_phone_number = input("Enter Phone Number: ").strip()

    for contact in numbers:
        if contact['phone_number'] == find_phone_number:
            print("\n===================================================")
            print("Contact found:")
            print(f"Name: {contact['first_name']} {contact['last_name']}")
            print(f"Phone Number: {contact['phone_number']}")
            print("===================================================")
            return

    print(f"No contact found with phone number: {find_phone_number}")


def find_contact_by_first_name():
    if not numbers:
        print("No contacts found.\n")
        return

    find_first_name = input("Enter First Name: ").strip()

    for contact in numbers:
        if contact['first_name'].lower() == find_first_name.lower():
            print("\n===================================================")
            print("Contact found:")
            print(f"Name: {contact['first_name']} {contact['last_name']}")
            print(f"Phone Number: {contact['phone_number']}")
            print("===================================================")
            return

    print(f"No contact found with first name: {find_first_name}")


def find_contact_by_last_name():
    if not numbers:
        print("No contacts found.\n")
        return

    find_last_name = input("Enter Last Name: ").strip()

    for contact in numbers:
        if contact['last_name'].lower() == find_last_name.lower():
            print("\n===================================================")
            print("Contact found:")
            print(f"Name: {contact['first_name']} {contact['last_name']}")
            print(f"Phone Number: {contact['phone_number']}")
            print("===================================================")
            return

    print(f"No contact found with last name: {find_last_name}")


def edit_contact():
    if not numbers:
        print("No contacts found.\n")
        return

    index_name = input("Please Enter the contact first name to update contact details: ").strip()

    for contact in numbers:
        if contact['first_name'].lower() == index_name.lower():
            contact['first_name'] = input("Enter the new contact first name: ").strip()
            contact['last_name'] = input("Enter the new contact last name: ").strip()
            contact['phone_number'] = input("Enter the new contact phone number: ").strip()
            print("The Contact has been updated successfully!!")
            return

    print("Contact not found")

File: daySeven/phoneBook/JavaScript/phoneBook.js
// phoneBook.js

const prompt = require('prompt-sync')();

let contacts = [];

function addContact() {
    let firstName;
    let lastName;
    let phoneNumber;

    while (true) {
        firstName = prompt("Please Enter First Name: ").trim();

        const duplicateFirstName = contacts.some(contact => contact.firstName.toLowerCase() === firstName.toLowerCase());
        if (duplicateFirstName) {
            console.log(`Error: User with the first name '${firstName}' already exists. Please enter a different name.`);
        } else {
            break;
        }
    }

    while (true) {
        lastName = prompt("Please Enter Last Name: ").trim();

        const duplicateLastName = contacts.some(contact => contact.lastName.toLowerCase() === lastName.toLowerCase());
        if (duplicateLastName) {
            console.log(`Error: User with the last name '${lastName}' already exists. Please enter a different name.`);
        } else {
            break;
        }
    }

    while (true) {
        phoneNumber = prompt("Enter Phone Number: ").trim();

        const duplicatePhoneNumber = contacts.some(contact => contact.phoneNumber === phoneNumber);
        if (duplicatePhoneNumber) {
            console.log(`Error: Phone number '${phoneNumber}' already exists. Please enter a different phone number.`);
        } else {
            break;
        }
    }

    const newContact = { firstName, lastName, phoneNumber };
    contacts.push(newContact);
    console.log("Contact has been added successfully!\n");
}

function removeContact() {
    if (contacts.length === 0) {
        console.log("No contacts found.\n");
        return;
    }

    const removePhoneNumber = prompt("Enter the phone number you want to remove: ").trim();
    const index = contacts.findIndex(contact => contact.phoneNumber === removePhoneNumber);

    if (index !== -1) {
        contacts.splice(index, 1);
        console.log("Contact has been removed successfully.");
    } else {
        console.log("Contact not found!!");
    }
}

function findContactByPhoneNumber() {
    if (contacts.length === 0) {
        console.log("No contacts found.\n");
        return;
    }

    const findPhoneNumber = prompt("Enter Phone Number: ").trim();
    const contact = contacts.find(contact => contact.phoneNumber === findPhoneNumber);

    if (contact) {
        console.log("\n===================================================");
        console.log(`Contact found:\nName: ${contact.firstName} ${contact.lastName}\nPhone Number: ${contact.phoneNumber}`);
        console.log("===================================================");
    } else {
        console.log(`No contact found with phone number: ${findPhoneNumber}`);
    }
}

function findContactByFirstName() {
    if (contacts.length === 0) {
        console.log("No contacts found.\n");
        return;
    }

    const findFirstName = prompt("Enter First Name: ").trim();
    const contact = contacts.find(contact => contact.firstName.toLowerCase() === findFirstName.toLowerCase());

    if (contact) {
        console.log("\n===================================================");
        console.log(`Contact found:\nName: ${contact.firstName} ${contact.lastName}\nPhone Number: ${contact.phoneNumber}`);
        console.log("===================================================");
    } else {
        console.log(`No contact found with first name: ${findFirstName}`);
    }
}

function findContactByLastName() {
    if (contacts.length === 0) {
        console.log("No contacts found.\n");
        return;
    }

    const findLastName = prompt("Enter Last Name: ").trim();
    const contact = contacts.find(contact => contact.lastName.toLowerCase() === findLastName.toLowerCase());

    if (contact) {
        console.log("\n===================================================");
        console.log(`Contact found:\nName: ${contact.firstName} ${contact.lastName}\nPhone Number: ${contact.phoneNumber}`);
        console.log("===================================================");
    } else {
        console.log(`No contact found with last name: ${findLastName}`);
    }
}

function editContact() {
    if (contacts.length === 0) {
        console.log("No contacts found.\n");
        return;
    }

    const indexName = prompt("Please Enter the contact first name to update contact details: ").trim();
    const contact = contacts.find(contact => contact.firstName.toLowerCase() === indexName.toLowerCase());

    if (!contact) {
        console.log("Contact not found");
        return;
    }

    contact.firstName = prompt("Enter the new contact first name: ").trim();
    contact.lastName = prompt("Enter the new contact last name: ").trim();
    contact.phoneNumber = prompt("Enter the new contact phone number: ").trim();

    console.log("The Contact has been updated successfully!!");
}

function mainMenu() {
    let userInput;
    do {
        console.log("\n===================================================");
        console.log("Welcome to PhoneBook");
        console.log("1. Add Contact");
        console.log("2. Remove Contact");
        console.log("3. Find contact by phone number");
        console.log("4. Find contact by first name");
        console.log("5. Find contact by last name");
        console.log("6. Edit contact");
        console.log("7. Exit");
        console.log("===================================================");

        userInput = parseInt(prompt("Choose an option: ").trim());

        switch (userInput) {
            case 1:
                addContact();
                break;
            case 2:
                removeContact();
                break;
            case 3:
                findContactByPhoneNumber();
                break;
            case 4:
                findContactByFirstName();
                break;
            case 5:
                findContactByLastName();
                break;
            case 6:
                editContact();
                break;
            case 7:
                console.log("Goodbye! Have a nice day!");
                break;
            default:
                console.log("Invalid Option. Please try again.");
        }
    } while (userInput !== 7);
}

mainMenu();
